### ISSUE TITLE
Subclasses of Twisted's protocol fail to be determined by mypy-zope

### ISSUE THREAD

**Issue by clokep** [Shoobx/mypy-zope#34]
Thanks for making mypy-zope! We've been using it for Synapse (which uses Twisted) and ran into a hiccup.

The following will cause mypy-zope to emit an error:

> error: Cannot determine consistent method resolution order (MRO) for "Foo"  [misc]

```python
from twisted.internet import protocol

class Foo(protocol.Protocol):
    pass

reveal_type(Foo)
```

The class hierarchy looks like `Foo` -> `Protocol` (which implements `IProtocol` and `ILoggingContext`) -> `BaseProtocol`.

I tried to create a reduced test-case for this, but was not able to, so it might be specific to something Twisted has done. Any help would be appreciated!

**Comment by kedder** [comment]
@clokep two questions:

1. Do you get this error when running mypy with cold caches (i.e. first time after you delete `.mypy_cache` directory?
2. Do you get this error if you are running mypy daemon? (i.e. by running `dmypy run`)

**Comment by clokep** [comment]
> Do you get this error when running mypy with cold caches (i.e. first time after you delete `.mypy_cache` directory?

Nope!

```sh
$ rm -r .mypy_cache
$ mypy temp.py
temp.py:6: note: Revealed type is 'def () -> temp.Foo'
$ mypy temp.py
temp.py:3: error: Cannot determine consistent method resolution order (MRO) for "Foo"  [misc]
temp.py:6: note: Revealed type is 'def () -> temp.Foo'
Found 1 error in 1 file (checked 1 source file)
```

> Do you get this error if you are running mypy daemon? (i.e. by running `dmypy run`)

I don't think so, but I've never used this before so I'm not 100% sure I'm using it properly.

**Comment by clokep** [comment]
I should also mention the versions of mypy I'm using:

mypy==0.812
mypy-extensions==0.4.3
mypy-zope==0.2.11

This is with the latest Twisted (Twisted==21.2.0).

**Comment by kedder** [comment]
Ok, then I think I'm observing the same problem in my project. We mainly use dmypy so this doesn't annoy us very much and we learned to ignore it. Maybe it is time to figure it out.

**Comment by clokep** [comment]
> Ok, then I think I'm observing the same problem in my project. We mainly use dmypy so this doesn't annoy us very much and we learned to ignore it.

I had completely forgotten dmypy existed until you mentioned it!

> Maybe it is time to figure it out.

I'd appreciate it. Please let me know if I can help at all! 😄

**Comment by clokep** [comment]
Any idea where one would start poking at to try to fix this? Does mypy-zope do something in particular to hook into the caching mechanism?

**Comment by kedder** [comment]
This might be a bit tricky. I think the reason we are getting this error is because of this hack: we are adding interface class to the MRO of all implementation classes in order for this assignment to work:
```
foo: IFoo = Foo()
```

mypy will consider `IFoo` as a supertype of `Foo` and therefore permit the assignment.

But at the same time, I believe having this interface in the MRO list confuses up the MRO calculation in some cases (makes the inheritance graph unsortable, or smth like that).

I'm not yet sure how to address that, but perhaps we need to find another way to support `foo: IFoo = Foo()` type of assignments.

**Comment by clokep** [comment]
This is hopefully fixed by #41, that at least fixes the example given in the description.
